### ISSUE TITLE
Won't start on Nexus 7

### ISSUE THREAD

**Issue by zubozrout** [pseuudonym404/neverball-touch#10]
Hi, Neverball won't start on Nexus 7 devel-proposed for the last 2 weeks or so. I would like to share any log here but I can't find one, nowhere.
Neverputt works just fine on the other hand.

Anyone has the same problem? Any idea what may be going on?

**Comment by laurie-bradshaw** [comment]
That is odd, due to the shared code between neverball and neverputt, I can't think why one would work while the other doesn't.

Neverball's own log should be in ~/.config/neverball.lb/neverball.log

stdout is also logged to ~/.cache/upstart/application-click-neverball.lb_neverball_1.6.0-4.log but that gets deleted when the application quits, so if it's crashing on startup it's going to be hard to grab that one. Touching the file and then tailing it before starting the app might work, I haven't tried.

**Comment by zubozrout** [comment]
I don't even have Neverball folder in ~/.config, only Neverputt. And I already tried reinstalling both the app and Ubuntu.

**Comment by laurie-bradshaw** [comment]
Could you try creating the directory (~/.config/neverball.lb) and see if you get a log?

Also, if that doesn't work, I just tested the touch and tail idea to get the stdout log and it works.

touch ~/.cache/upstart/application-click-neverball.lb_neverball_1.6.0-4.log
tail -f ~/.cache/upstart/application-click-neverball.lb_neverball_1.6.0-4.log

Then launch the app.

**Comment by zubozrout** [comment]
Hi. Creating directory didn't help but the second tip worked. Here it is:

neverball: error while loading shared libraries: libubuntu_application_api.so.2: cannot open shared object file: No such file or directory

**Comment by zubozrout** [comment]
On devel-proposed it seems this library is missing and only so.3 is present.

**Comment by laurie-bradshaw** [comment]
Could you try putting libubuntu_application_api.so.2 from the current release in /opt/click.ubuntu.com/neverball.lb/1.6.0-4/lib/arm-linux-gnueabihf

I can put the copy from my phone on dropbox or the like if you don't have it handy.

If that works, I'll bundle it for now, and rebuild against a newer sdk when devel-proposed becomes the release version.

**Comment by zubozrout** [comment]
Hmm, apparently there are more newer libraries not available in their older versions, including libubuntu_application_api.so.2, libmirclient.so.8 and libmircommon.so.2.

**Comment by laurie-bradshaw** [comment]
I'll try rebuilding sdl with dynamic loading of the mir libraries.

Dynamically loading libubuntu_application_api shold be easy, neverball only uses a couple of functions for the tilt sensor.

It's apparently possible to keep the backlight on now, so I'm also trying to figure out how to do that without qt, so it might take a little while.

It'd be helpful if you could post the output of

ldd /opt/click.ubuntu.com/neverball.lb/1.6.0-4/lib/arm-linux-gnueabihf/bin/neverball

and

ldd /opt/click.ubuntu.com/neverball.lb/1.6.0-4/lib/arm-linux-gnueabihf/libSDL2-2.0.so.0

**Comment by zubozrout** [comment]
Hi,
thank you very much for the feedback. Here are the outputs for not-patched standard devel-proposed:

ldd /opt/click.ubuntu.com/neverball.lb/1.6.0-4/lib/arm-linux-gnueabihf/bin/neverball:
    libubuntu_application_api.so.2 => not found
    libSDL2_ttf-2.0.so.0 => not found
    libvorbisfile.so.3 => not found
    libSDL2-2.0.so.0 => not found
    libpthread.so.0 => /lib/arm-linux-gnueabihf/libpthread.so.0 (0xb6f6a000)
    libGLESv1_CM.so.1 => not found
    libjpeg.so.8 => /usr/lib/arm-linux-gnueabihf/libjpeg.so.8 (0xb6f2a000)
    libpng12.so.0 => /lib/arm-linux-gnueabihf/libpng12.so.0 (0xb6f02000)
    libm.so.6 => /lib/arm-linux-gnueabihf/libm.so.6 (0xb6e8d000)
    libc.so.6 => /lib/arm-linux-gnueabihf/libc.so.6 (0xb6d9f000)
    /lib/ld-linux-armhf.so.3 (0xb6f91000)
    libz.so.1 => /lib/arm-linux-gnueabihf/libz.so.1 (0xb6d7c000)

ldd /opt/click.ubuntu.com/neverball.lb/1.6.0-4/lib/arm-linux-gnueabihf/libSDL2-2.0.so.0:
    libm.so.6 => /lib/arm-linux-gnueabihf/libm.so.6 (0xb6e43000)
    libdl.so.2 => /lib/arm-linux-gnueabihf/libdl.so.2 (0xb6e30000)
    libpulse.so.0 => /usr/lib/arm-linux-gnueabihf/libpulse.so.0 (0xb6ded000)
    libmirclient.so.8 => not found
    libxkbcommon.so.0 => /usr/lib/arm-linux-gnueabihf/libxkbcommon.so.0 (0xb6dae000)
    libpthread.so.0 => /lib/arm-linux-gnueabihf/libpthread.so.0 (0xb6d99000)
    librt.so.1 => /lib/arm-linux-gnueabihf/librt.so.1 (0xb6d83000)
    libc.so.6 => /lib/arm-linux-gnueabihf/libc.so.6 (0xb6c95000)
    /lib/ld-linux-armhf.so.3 (0xb6f70000)
    libjson-c.so.2 => /lib/arm-linux-gnueabihf/libjson-c.so.2 (0xb6c7e000)
    libpulsecommon-6.0.so => /usr/lib/arm-linux-gnueabihf/pulseaudio/libpulsecommon-6.0.so (0xb6c1f000)
    libdbus-1.so.3 => /lib/arm-linux-gnueabihf/libdbus-1.so.3 (0xb6be3000)
    libxcb.so.1 => /usr/lib/arm-linux-gnueabihf/libxcb.so.1 (0xb6bbf000)
    libsystemd.so.0 => /lib/arm-linux-gnueabihf/libsystemd.so.0 (0xb6b6b000)
    libwrap.so.0 => /lib/arm-linux-gnueabihf/libwrap.so.0 (0xb6b5d000)
    libsndfile.so.1 => /usr/lib/arm-linux-gnueabihf/libsndfile.so.1 (0xb6b03000)
    libasyncns.so.0 => /usr/lib/arm-linux-gnueabihf/libasyncns.so.0 (0xb6aee000)
    libXau.so.6 => /usr/lib/arm-linux-gnueabihf/libXau.so.6 (0xb6ae3000)
    libXdmcp.so.6 => /usr/lib/arm-linux-gnueabihf/libXdmcp.so.6 (0xb6acf000)
    liblzma.so.5 => /lib/arm-linux-gnueabihf/liblzma.so.5 (0xb6aaf000)
    libgcrypt.so.20 => /lib/arm-linux-gnueabihf/libgcrypt.so.20 (0xb6a20000)
    libgcc_s.so.1 => /lib/arm-linux-gnueabihf/libgcc_s.so.1 (0xb69f5000)
    libnsl.so.1 => /lib/arm-linux-gnueabihf/libnsl.so.1 (0xb69d5000)
    libFLAC.so.8 => /usr/lib/arm-linux-gnueabihf/libFLAC.so.8 (0xb6989000)
    libvorbisenc.so.2 => /usr/lib/arm-linux-gnueabihf/libvorbisenc.so.2 (0xb68f7000)
    libresolv.so.2 => /lib/arm-linux-gnueabihf/libresolv.so.2 (0xb68d7000)
    libgpg-error.so.0 => /lib/arm-linux-gnueabihf/libgpg-error.so.0 (0xb68ba000)
    libogg.so.0 => /usr/lib/arm-linux-gnueabihf/libogg.so.0 (0xb68ad000)
    libvorbis.so.0 => /usr/lib/arm-linux-gnueabihf/libvorbis.so.0 (0xb687b000)

**Comment by laurie-bradshaw** [comment]
Thanks. :)

I'd forgotten you'd said neverputt is working, rebuilding sdl probably isn't necessary, it's probably just libubuntu_application_api for the tilt sensor in neverball, which neverputt doesn't use, the bundled sdl library in both is the same binary.

Should be a straightforward fix, I'll get to it as soon as I figure out the backlight.

**Comment by laurie-bradshaw** [comment]
I just uploaded a version that dynamically loads ubuntu_application_api (and keeps the backlight on), if you could let me know if that worked? :)

**Comment by zubozrout** [comment]
Hmm, sadly this didn't help. Neverball's log at ~/.cache/upstart/application-click-neverball.lb_neverball_1.6.0-5.log now contains this:
neverball: error while loading shared libraries: libmirclient.so.8: cannot open shared object file: No such file or directory
Neverputt now stopped working and I assume there is the same problem with it although I only get empty log when trying to retrieve it ... maybe I am using a wrong filename for it.

**Comment by laurie-bradshaw** [comment]
For some reason, neverputt's stdout log is .cache/upstart/application-click-neverputt.lb_neverball_1.6.0-6.log

I guess SDL does need to load mir dynamically.

**Comment by zubozrout** [comment]
Thanks :). Yes, it is the same library missing.

**Comment by laurie-bradshaw** [comment]
Seems dynamic loading of mir is broken in current SDL, and wouldn't help with the different versions anyway.

Let's try hacking it instead, as it's only a temporary workaround until devel-proposed becomes stable, it can just be rebuilt against the newer libs then.

could you try doing

sudo ln -s /usr/lib/arm-linux-gnueabihf/libmirclient.so.9 /opt/click.ubuntu.com/neverball.lb/1.6.0-5/lib/arm-linux-gnueabihf/libmirclient.so.8

and see if that gets it running?

**Comment by zubozrout** [comment]
Won't start:
neverball: /usr/lib/arm-linux-gnueabihf/libmirclient.so.8: version 'MIR_CLIENT_8' not found (required by /opt/click.ubuntu.com/.click/users/phablet/neverball.lb/lib/arm-linux-gnueabihf/libSDL2-2.0.so.0)

**Comment by laurie-bradshaw** [comment]
Do you have a copy of the current libmirclient.so.8 to put in /opt/click.ubuntu.com/neverball.lb/1.6.0-5/lib/arm-linux-gnueabihf instead of the symlink?

**Comment by zubozrout** [comment]
Yes, fortunately I have. But it still crashes - although log files for both Neverball and Neverputt are empty in this case.

**Comment by laurie-bradshaw** [comment]
It's looking like I can't make the package on the store work on both stable and devel-proposed at the same time.

I can do a build of sdl for devel-proposed that you can replace the bundled one with that should get it working though.

**Comment by laurie-bradshaw** [comment]
Actually, I can use a startup script to check which version of libmirclient is available and set the library path for sdl accordingly if this sdl build works (just setting up the devel-proposed chroot now).

**Comment by laurie-bradshaw** [comment]
click chroot -a armhf -f ubuntu-sdk-15.10 create is broken too. :/

It installed ok, but doesn't work and can't be removed.

**Comment by zubozrout** [comment]
Lol. Well, devel-proposed isn't recommended for users so I think it is ok for you to ignore it for now. devel-proposed won't be even usable once they'll be switching to snappy packages (I have no Idea when is this planned).

**Comment by laurie-bradshaw** [comment]
Yeah, devel-proposed seems pretty much broken for 3rd party apps atm, probably best to ignore it until it's clear how software is supposed to be built for it. :)

I'll leave this open though, as it does want fixing at some point.

**Comment by zubozrout** [comment]
Thanks for the info. Unfortunately on Nexus 7 LTE only devel-proposed makes sense. Yesterday I tried flo image of rc-proposed on my deb device but it had several problems - no sound, stacked in landscape mode (landcape apps were rotated to portrait :) ), so I reverted back to deb's devel-proppsed image. And sadly there was no update on devel-proposed for flo/deb for quite some time now - I wonder what is going on there.

**Comment by laurie-bradshaw** [comment]
I've managed to build SDL for 15.10. I had to tweak the video code a bit and completely rewrite the input code due to changes in mir, and I have nothing running 15.10 (yet), so it's completely untested.

https://www.dropbox.com/s/ag711szy0ltbec7/libSDL2-2.0.so.0
edit: patch against current sdl mercurial for above build: https://www.dropbox.com/s/77850ikrsvayri7/sdl-mirchanges.patch

Assuming I didn't get anything wrong, replacing libSDL with this copy should get neverputt working. current neverball release is linked against the older libubuntu_application_api.

Gettting something running 15.10 (emulator?) so I can actually test is on my todo list. :)

**Comment by zubozrout** [comment]
Hmm, replacing that didn't help for me ... but maybe I am doing something wrong. What I did right however was purchasing an official Ubuntu phone ... so I can play your games :+1: :)

Thank you very much for your support.

**Comment by laurie-bradshaw** [comment]
Heh, I'm trying to do the opposite; find a cheap nexus to run devel on for testing.

And no problem, I guess everything will run 15.10 eventually, so it needs fixing. :)

**Comment by zubozrout** [comment]
Hmm ... not planning to get rid of mine Nexus 7 yet. Isn't Ubuntu emulator enough for testing?

**Comment by laurie-bradshaw** [comment]
I'm watching a couple on ebay, but thanks for the thought.

The emulator is great, but real hardware is better.

I'd get a meizu if I could afford it, and run devel on the bq. :)

**Comment by zubozrout** [comment]
Yup, I feel with you :'(. I am a student so I know what you are talking about :'(, although I picked Meizu as I really liked the camera and would miss LTE. But also it is a substantial upgrade after so many years :).

Hope you'll have luck with your search.

In the meantime ... If you have some time to spare ... It would be great to read about your development of all those SDL games for Ubuntu Phone ... There is very little on this topic so some quicky tutorial would be awesome for Ubuntu Phone :).

**Comment by laurie-bradshaw** [comment]
It essentially amounts to just using click. :)
https://developer.ubuntu.com/en/blog/2015/04/13/inner-workings-sdk/

I'm planning on putting together a mini sdl sdk with libraries, headers and demo apps (sdl tests) once I've got sdl working on 15.10.

**Comment by zubozrout** [comment]
Finally got some time to at least try it and yup ... on 15.10 we are out of luck:
"warning: libmirclient.so.8, needed by lib/arm-linux-gnueabihf/libSDL2.so, not found (try using -rpath or -rpath-link)"

I'll try to get to try 15.04 at the weekend :).
